File: Users/jeong-yeongjae/sub_tree_test/iOS/FitGAI/PoseLogic/Classification/ClassificationResult.swift
//
//  ClassificationResult.swift
//  FitGAI
//
//  Created by 김기훈 on 2021/12/13.
//

import Foundation

class ClassificationResult {
    var classConfidences:[String:Float]
    init() {
        classConfidences = [:]
    }
    
    func getAllClasses() -> Set<String> {
        var set = Set<String>()
        for key in classConfidences.keys {
            set.insert(key)
        }
        return set
    }
    
    func getClassConfidence(className: String) -> Float {
        if classConfidences[className] != nil {
            return classConfidences[className]!
        }  else  {
            return 0.0
        }
     }
    
//    func getMaxConfidenceClass() -> String {
//        let sortClassConfidences = classConfidences.sorted { (left: Dictionary<String, Float>.Element, right:  Dictionary<String, Float>.Element) -> Bool in
//            return left.value < right.value
//        }
//        return sortClassConfidences.last?.key ?? "down"
//    }
    
    func getMaxConfidenceClass() -> String {
        let maxValue = classConfidences.values.max()
        for item in classConfidences {
            if item.value == maxValue {
                return item.key
            }
        }
        return "ERROR"
    }
    
    func incrementClassConfidence(className: String?) {
        if (classConfidences[className!] != nil )  {
            classConfidences[className!] = classConfidences[className!]! + 1
        }
        else{
            classConfidences[className!] = 1
        }
    }
    func putClassConfidence(className: String, confidence: Float) {
        classConfidences[className] = confidence
    }

}

File: Users/jeong-yeongjae/sub_tree_test/presentation/src/main/java/com/hnh/presentation/logic/classification/ClassificationResult.kt
package com.hnh.presentation.logic.classification

import java.util.*

class ClassificationResult {
    // For an entry in this map, the key is the class name, and the value is how many times this class
    // appears in the top K nearest neighbors. The value is in range [0, K] and could be a float after
    // EMA smoothing. We use this number to represent the confidence of a pose being in this class.
    private val classConfidences: MutableMap<String, Float>
    fun getAllClasses(): Set<String> {
        return classConfidences.keys
    }
    fun getClassConfidence(className: String): Float {
        return if (classConfidences.containsKey(className)) classConfidences[className]!! else 0F
    }

    fun getMaxConfidenceClass(): String {
        return Collections.max<Map.Entry<String, Float>>(
            classConfidences.entries
        ) { entry1: Map.Entry<String, Float>, entry2: Map.Entry<String, Float> -> (entry1.value - entry2.value).toInt() }
            .key
    }

    fun incrementClassConfidence(className: String?) {
        classConfidences[className!!] =
            if (classConfidences.containsKey(className)) classConfidences[className]!! + 1F else 1F
    }

    fun putClassConfidence(className: String, confidence: Float) {
        classConfidences[className] = confidence
    }

    init {
        classConfidences = HashMap()
    }
}
